### ISSUE TITLE
dll not found

### ISSUE THREAD

**Issue by mfrank29** [xfleckx/LSL4Unity#15]
Hi, 
I get this error on start:
DllNotFoundException: Assets/Extensions/LSL4Unity/lib/liblsl64.dll
I simply followed the instructions from the wiki. I added the package and add a LSLMarkerStream Component on a GameObject. Then I use the marker stream just like in the instructions. 
Did I miss something or do I have to consider something while importing the package?

**Comment by xfleckx** [comment]
I'm sorry, 
looks like something got wrong with the upgrade to LSL version 1.12 - I will take a look as soon as possible!

**Comment by cboulay** [comment]
I did add the new binaries in 85f0a8ac05849f53ccdb7580c21dbaa35340511e .

@mfrank29 , do you have the dll anywhere in your LSL4Unity tree? If yes, where is it?

**Comment by mfrank29** [comment]
@cboulay: Yes, I have. The Path is /UnityProject/Assets/Extensions/LSL4Unity/lib/liblsl64.dll

**Comment by cboulay** [comment]
I think you need to copy paste more of the error message. Do you have any "unresolved symbol" or similar?

**Comment by mfrank29** [comment]
This is the full Error message:
DllNotFoundException: Assets/Extensions/LSL4Unity/lib/liblsl64.dll
LSL.liblsl+StreamInfo..ctor (System.String name, System.String type, Int32 channel_count, Double nominal_srate, channel_format_t channel_format, System.String source_id) (at Assets/Extensions/LSL4Unity/LSL.cs:151)
Assets.LSL4Unity.Scripts.LSLMarkerStream.Awake () (at Assets/Extensions/LSL4Unity/Scripts/LSLMarkerStream.cs:30)

**Comment by xfleckx** [comment]
Hy I'm sorry the problem but is the _Extension_ subdirectory!

please move the LSL4Unity asset package to the Asset directory without another subdirectory!

**Comment by xfleckx** [comment]
Hy, 

I changed the all necessary things to use the library no as Unity intent to use native plugins. With the latest [commit](https://github.com/xfleckx/LSL4Unity/commit/3f83ab6a0ad545c76de870c18c07d71f24bb2aba) it's possible to put the plugin in any subdirectory beneath Assets.

**Comment by mfrank29** [comment]
Hi,
The error still remains. I moved the directory to the Assets folder before, to test if this was the mistake, but there was no difference. In the current version the .dll files in the plugin folder were missing, so I took the one s from an older version. These are the Errors that were showing up during runtime:

Failed to load 'Assets/Extensions/LSL4Unity/Plugins/liblsl64.dll' with error 'Das angegebene Modul wurde nicht gefunden.
'.
Assets.LSL4Unity.Scripts.LSLMarkerStream:Awake() (at Assets/Extensions/LSL4Unity/Scripts/LSLMarkerStream.cs:39)
Assets.LSL4Unity.Scripts.LSLMarkerStream:Awake() (at Assets/Extensions/LSL4Unity/Scripts/LSLMarkerStream.cs:30)
Failed to load 'Assets/Extensions/LSL4Unity/Plugins/liblsl64.dll' with error 'Das angegebene Modul wurde nicht gefunden.
'.
LSL.StreamInfo:.ctor(String, String, Int32, Double, channel_format_t, String) (at Assets/Extensions/LSL4Unity/LSL.cs:151)
LSL.StreamInfo:.ctor(String, String, Int32, Double, channel_format_t, String) (at Assets/Extensions/LSL4Unity/LSL.cs:151)
Assets.LSL4Unity.Scripts.LSLMarkerStream:Awake() (at Assets/Extensions/LSL4Unity/Scripts/LSLMarkerStream.cs:30)

DllNotFoundException: Assets/Extensions/LSL4Unity/Plugins/liblsl64.dll
LSL.liblsl+StreamInfo..ctor (System.String name, System.String type, Int32 channel_count, Double nominal_srate, channel_format_t channel_format, System.String source_id) (at Assets/Extensions/LSL4Unity/LSL.cs:151)
Assets.LSL4Unity.Scripts.LSLMarkerStream.Awake () (at Assets/Extensions/LSL4Unity/Scripts/LSLMarkerStream.cs:30)

Could there be something wrong with the .dll files?

**Comment by cboulay** [comment]
You can't use old versions of the dll files because they don't have the functions that the code is looking for. Download the latest liblsl-C-C++-<version>.zip from [here](ftp://sccn.ucsd.edu/pub/software/LSL/SDK/).

**Comment by xfleckx** [comment]
I'm sorry guys....

I forgot to add the dll's (since dlls should not be part of a repository we need to make an exception)...

Will fix that todays evening

**Comment by xfleckx** [comment]
hy, 
so last commit added the missing files... sorry for the inconvenience :/ hope that it will work as expected :)

**Comment by mfrank29** [comment]
Hi, 
I'm sorry, but theres is no difference. The Error messages are still showing up at start. :/

**Comment by xfleckx** [comment]
I'm sorry... again... 
the dll from the LSL build seems to be missing a dependency - which could not be resolved by Unitys application context... The one who build the library had used a wrong build configuration... (maybe wrong target version of the windows runtime... )

![image](https://cloud.githubusercontent.com/assets/1039313/24463316/bd326cdc-14a6-11e7-9c11-55efcaa14761.png)

I will add the older (but still working well - libraries 1.11) todays evening... 

Actually I've no idea why it was working on my machine.... maybe

**Comment by cboulay** [comment]
The old libraries might also raise an error because there are some new functions or function signatures that are referenced by the new C# wrapper.

Where did you get the current version of the dll? From me or from the ftp? I can try to rebuild them all. I used self-built on my machine and used those without issue for an experiment. But it's also possible that I have an MSVC runtime installed on my PC that mfrank29 does not have.

@mfrank29 , can you try installing the [MSVC2015 redistributable](https://www.microsoft.com/en-us/download/details.aspx?id=48145)?

**Comment by mfrank29** [comment]
@cboulay
I get it from the newest commit here on github. I tried to install the MSVC2015 redistributable, but it cannot be installed, because a newer version is installed

**Comment by xfleckx** [comment]
Unfortunately, I tried them from the ftp... @cboulay Thanks for your help, could you try the current HEAD in a blank project? just to make sure whether it is working on your machine?

**Comment by cboulay** [comment]
Sorry I wasn't able to get to it today. It might not be possible for the rest of the week. I will update you if I get a chance to test it.

**Comment by cboulay** [comment]
I remote'd in to my Windows box to give a quick try. I created a blank project. I closed the project, opened command prompt to its Assets folder, and `git clone`d this repo.

Back in the Editor, I added a cube to the scene then attached the LSLMarkerStream script to it. I press play and I can see the new stream on the network. When I "unpress" play, Unity crashes.

This is the same for x86 or x86_64 targets.

This machine isn't really good for development or debugging. I will try to come back to this next week when I'm back in the lab.

**Comment by xfleckx** [comment]
Hy,

I tried it again on my development machine - the HEAD (latest commit) is working properly without crashes. I hate my self for serving that "it works for me" cliché.

However, it's very likely that the reason is somewhere in the redistributables. I make an educated guess... It's the MSVCR**90**.dll and everything which got shipped with the VC 2008 Redistributable... pretty old that thing... 

![image](https://cloud.githubusercontent.com/assets/1039313/24491693/b1ac1d2c-1528-11e7-933a-da16de61fb50.png) 

Another guess: the other missing things could be ignored [since the dependencyWalker could not handle the API redirection introduced with Windows 7](https://ofekshilon.com/2016/03/27/on-api-ms-win-xxxxx-dll-and-other-dependency-walker-glitches/)

**Possible solution for** @mfrank29 
 1. get a cup of coffee that will take a while
 2. [check this page](http://asawicki.info/news_1597_installing_visual_c_redistributable_package_from_command_line.html) for the command line calls enabling the installation of the redistributables packages - (start with 2008) even if newer versions already installed.. I guess the silent option will do the job... so many guesses... I'm glad that software engineers don't build bridges...

**Comment by xfleckx** [comment]
Unfortunately, it's not the only problem :( I tried it on a machine where the expected missing redistributables are installed and got the same issue... I will make a new branch for further investigations and return to the lib version 1.11 for now so that it @mfrank29 got a stable version...

**Comment by mfrank29** [comment]
I've installed the redistributables, or rather they were installed:
![installedvs](https://cloud.githubusercontent.com/assets/3041779/24496398/776bccd8-1538-11e7-8dd0-f723d1986d1d.png)
But like you already said, that wasn't the only problem... al least i got coffee ^^
So should I download an older version, which have the 1.11 libaries?

**Comment by xfleckx** [comment]
I prepared the older version for you in a new branch.. see the previous message... :)

**Comment by mfrank29** [comment]
There are no Error messages anymore. I think it will work now, thank you very much :)

**Comment by otraupe** [comment]
Markus, thanks again for your great support!

**Comment by xfleckx** [comment]
Glad to hear that you could now work with the thing :man_factory_worker: 

This issue made me trying to refactor the package to the more "Unitysh" way for handling native plugins, which is a little improvement. 👍 😊 

Thanks for using LSL4Unity and contributing using issue requests!

**Comment by Tinuxx** [comment]
Hi,
I'm having this exact same issue when trying to run the demo scenes on a machine running OS X. I follow the steps listed on the wiki: download repo as .zip, and move directory under /Assets/. Then, when I hit Run, this comes up in the console:

Demo_IncomingStreams:
```
DllNotFoundException: liblsl64.bundle
LSL.liblsl+ContinuousResolver..ctor (Double forget_after) (at Assets/LSL4Unity/LSL.cs:837)
Assets.LSL4Unity.Scripts.Resolver.Start () (at Assets/LSL4Unity/Scripts/Resolver.cs:35)
```

Demo_RotatingCube (preceded by several errors that are identical to the previous one):
```
NullReferenceException: Object reference not set to an instance of an object
Assets.LSL4Unity.Scripts.LSLMarkerStream.Write (System.String marker) (at Assets/LSL4Unity/Scripts/LSLMarkerStream.cs:44)
RandomMarker+<WriteContinouslyMarkerEachSecond>c__Iterator0.MoveNext () (at Assets/LSL4Unity/Demos/RandomMarker.cs:27)
UnityEngine.SetupCoroutine.InvokeMoveNext (IEnumerator enumerator, IntPtr returnValueAddress) (at /Users/builduser/buildslave/unity/build/Runtime/Export/Coroutines.cs:17)
UnityEngine.MonoBehaviour:StartCoroutine(IEnumerator)
RandomMarker:Start() (at Assets/LSL4Unity/Demos/RandomMarker.cs:18)
```

Do I need to do any further configuration (say, in the Plugins inspector)? I'm really stumped, as the liblsl64.bundle file is clearly there... any suggestions? Many thanks!

**Comment by Tinuxx** [comment]
I just stumbled into the fix. On [this line](https://github.com/xfleckx/LSL4Unity/blob/a0d75ffc0b711064f6396fb3b94ad795a12bf649/LSL.cs#L947), I replaced `"liblsl64.bundle"` with `"liblsl64"`. Just in case anyone is having the same issue on Mac.

Many thanks for making this project! 😄

**Comment by mariusrubo** [comment]
As described in the linked issue above, I can't get LSL4Unity to work on my 64Bit Windows 7 computer. 
I downloaded the LSL4Unity master branch, copied it in my newly created Unity project, dragged the LabStreamingLayer.prefab in the scene and pressed play. 
I got these error messages:

> Failed to load 'Assets/LSL4Unity/Plugins/liblsl64.dll' with error 'Das angegebene Modul wurde nicht gefunden.
'.
Assets.LSL4Unity.Scripts.LSLMarkerStream:Awake() (at Assets/LSL4Unity/Scripts/LSLMarkerStream.cs:39)
Assets.LSL4Unity.Scripts.LSLMarkerStream:Awake() (at Assets/LSL4Unity/Scripts/LSLMarkerStream.cs:30)

and 

> DllNotFoundException: Assets/LSL4Unity/Plugins/liblsl64.dll
LSL.liblsl+StreamInfo..ctor (System.String name, System.String type, Int32 channel_count, Double nominal_srate, channel_format_t channel_format, System.String source_id) (at Assets/LSL4Unity/LSL.cs:151)
Assets.LSL4Unity.Scripts.LSLMarkerStream.Awake () (at Assets/LSL4Unity/Scripts/LSLMarkerStream.cs:30)


liblsl64.dll is however indeed located in the folder where Unity claims it cannot find it. Any ideas what I could do?

**Comment by xfleckx** [comment]
Which branch are you using? Plz. Try the using_lib111 branch instead of master.

**Comment by mariusrubo** [comment]
Oh I see, I should have read your previous comments more carefully ... ;-) Sorry about that. 
There are no more error messages with lib111. I'll now try to fetch a stream from our OpenBCI system.

**Comment by cboulay** [comment]
I just tried to use LSL4Unity for the first time in a long while. Just like the users above, it failed to load the liblsl64.dll.

There have been many changes to the LSL build system in recent months, and I'm not sure exactly what the important difference is, but I dropped in the new DLL and it just worked. I also had to change the line that loads the DLL because the filename has changed (now lsl64_MSVC.dll). Note that these new DLLs are NOT on the ftp.

**Comment by cboulay** [comment]
Oops, I forgot that the new filename was just an intermediate step and the newest filename now matches the older filenames (i.e., liblsl64.dll). I've updated my previous post (the new dll with the old filename still works).

**Comment by mvidaldp** [comment]
Hi,
I'm getting this same error when trying to run my Unity Project with LSL4Unity on Linux (actually on Manjaro). If I run the same code on Windows it works without any problems. The error message is next:

DllNotFoundException: liblsl64.so
LSL4Unity.liblsl+StreamInfo..ctor (System.String name, System.String type, System.Int32 channel_count, System.Double nominal_srate, LSL4Unity.liblsl+channel_format_t channel_format, System.String source_id) (at Assets/LSL4Unity/LSL.cs:195)
DataStreams.Start () (at Assets/Scripts/DataStreams.cs:56)

The error line on LSL.cs is next:
obj = dll.lsl_create_streaminfo(name, type, channel_count, nominal_srate, channel_format, source_id); 

On DataStream.cs is next:
_lslStreamInfo = new liblsl.StreamInfo(
                t.sName,
                t.sType.ToString(),
                lslChannelCount,
                _nominalRate,
                LslChannelFormat,
                uuid.ToString());

Any ideas about how to fix it? I tried using different DLLs. For instance, liblsl64.dll and also renaming the Linux one on LSL.cs without the .so extension as it was an issue on macOS. I also tried using liblsl from the package installed using my package manager and also the compiled one from the cloned repo. I also tried a different folder structure but I always get the same error.
